examples/stream/src/lib.rs

```diff
@@ -0,0 +1,517 @@
+
+    // Note: Connections will perform work-stealing on answering `NewStream` messages.
+    connections_by_peer_id: HashMap<PeerId, (flume::Sender<NewStream>, flume::Receiver<NewStream>)>,
+
+    events: VecDeque<ToSwarm<(), ToHandler>>,
+
+    pending_connections: HashMap<
+        PeerId,
+        Vec<oneshot::Sender<io::Result<flume::r#async::SendSink<'static, NewStream>>>>,
+    >,
+}
+
+pub struct Handler {
+    remote: PeerId,
+    supported_protocols: HashMap<StreamProtocol, mpsc::Sender<(PeerId, Stream)>>,
+
+    receiver: flume::r#async::RecvStream<'static, NewStream>,
+    pending_upgrade: Option<(StreamProtocol, oneshot::Sender<Result<Stream, Error>>)>,
+}
+
+impl Handler {
+    fn new(
+        remote: PeerId,
+        supported_protocols: HashMap<StreamProtocol, mpsc::Sender<(PeerId, Stream)>>,
+        receiver: flume::r#async::RecvStream<'static, NewStream>,
+    ) -> Self {
+        Self {
+            supported_protocols,
+            receiver,
+            pending_upgrade: None,
+            remote,
+        }
+    }
+}
+
+impl Default for Behaviour {
+    fn default() -> Self {
+        let (sender, receiver) = mpsc::channel(0);
+
+        Self {
+            sender,
+            receiver,
+            connections_by_peer_id: HashMap::default(),
+            active_connections: HashSet::default(),
+            events: VecDeque::default(),
+            supported_protocols: HashMap::default(),
+            pending_connections: HashMap::default(),
+        }
+    }
+}
+
+impl Behaviour {
+    pub fn register(
+        &mut self,
+        protocol: StreamProtocol,
+    ) -> Result<(Control, IncomingStreams), AlreadyRegistered> {
+        if self.supported_protocols.contains_key(&protocol) {
+            return Err(AlreadyRegistered);
+        }
+
+        let (sender, receiver) = mpsc::channel(10);
+        self.supported_protocols
+            .insert(protocol.clone(), sender.clone());
+
+        self.events
+            .extend(
+                self.active_connections
+                    .iter()
+                    .map(|(peer, conn)| ToSwarm::NotifyHandler {
+                        peer_id: *peer,
+                        handler: swarm::NotifyHandler::One(*conn),
+                        event: ToHandler::RegisterProtocol(RegisterProtocol {
+                            protocol: protocol.clone(),
+                            sender: sender.clone(),
+                        }),
+                    }),
+            );
+
+        tracing::debug!(
+            %protocol,
+            "Registering protocol with {} existing handlers",
+            self.active_connections.len()
+        );
+
+        Ok((
+            Control {
+                sender: self.sender.clone(),
+                protocol: protocol,
+            },
+            IncomingStreams { receiver },
+        ))
+    }
+}
+
+#[derive(Debug)]
+pub struct AlreadyRegistered;
+
+// TODO: On `Drop`, we need to de-register the protocol.
```

Comment:
On drop of all instances of `Control`. Correct?

Though given that this carries a `mpsc::Sender`, one might as well do the garbage collection when the `Receiver` end closes.

Reply:
I just thought about this some more and we don't actually need to do this. The `Control` is only for _opening_ new streams. It is a perfectly legit use of this module to immediately drop the `Sender` if your application is never going to open a stream for a certain protocol. Plus, I actually think you should be able to obtain a new `Control` at any point.

What should de-register the protocol is dropping `IncomingStreams`. Thinking more about this, it seems like a better API to decouple these two things. One should be able to obtain a `Control` without being forced to accept inbound streams using the same one.

Reply:
> One should be able to obtain a `Control` without being forced to accept inbound streams using the same one.

Reading your first paragraph, I had the same thought.

Maybe separate methods for all 3 options (in, out, in and out)? Or an `enum` as a parameter to the method?

Reply:
> > One should be able to obtain a `Control` without being forced to accept inbound streams using the same one.
> 
> Reading your first paragraph, I had the same thought.
> 
> Maybe separate methods for all 3 options (in, out, in and out)? Or an `enum` as a parameter to the method?

Have a look at the latest version. I quite like the simplicity of a dedicated `new_control` and a `accept` function. The names could be improved. Do you have any ideas?